server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletTransactionLog.java

```diff
@@ -0,0 +1,388 @@
+      transaction.apply(newFiles);
+      return newFiles.toArray(new StoredTabletFile[0]);
+    }
+
+    Date getInitialDate() {
+      return Date.from(Instant.ofEpochMilli(initialTs));
+    }
+
+    int getNumTransactions() {
+      return tabletLog.size();
+    }
+
+    List<TabletTransaction> getTransactions() {
+      return tabletLog.toList();
+    }
+
+    Set<StoredTabletFile> getExpectedFiles() {
+      return new HashSet<>(Arrays.asList(finalFiles));
+    }
+
+    boolean isEmpty() {
+      return tabletLog.isEmpty();
+    }
+
+    /**
+     * Return a human readable dump of the log
+     *
+     * @param extent The tablet extent
+     * @return A dump of the log
+     */
+    String dumpLog(KeyExtent extent, boolean clear) {
+      // first lets get a consistent set of files and transactions
+      long initialTs;
+      StoredTabletFile[] initialFiles;
+      List<TabletTransaction> transactions;
+      StoredTabletFile[] finalFiles;
+      boolean consistent;
+      do {
+        long updateCount = this.updateCount;
+        initialTs = this.initialTs;
+        initialFiles = this.initialFiles;
+        finalFiles = this.finalFiles;
+        transactions = this.tabletLog.toList();
+        consistent =
+            (updateCount == this.updateCount && updateCount == this.tabletLog.getUpdateCount());
+      } while (!consistent);
+
+      // now clear out the log if requested
+      if (clear) {
+        clear();
+      }
+
+      // now we can build our dump string
+      StringBuilder builder = new StringBuilder();
+      SimpleDateFormat format = new SimpleDateFormat(DATE_FORMAT);
+      String initialDate = format.format(Date.from(Instant.ofEpochMilli(initialTs)));
+
+      builder.append(String.format("\n%s: Initial files for %s : %s", initialDate, extent,
+          new TreeSet<>(Arrays.asList(initialFiles))));
+      transactions.stream().forEach(t -> builder.append('\n').append(t.toString(format)));
+      if (!transactions.isEmpty()) {
+        builder.append("\nFinal files: ").append(new TreeSet<>(Arrays.asList(finalFiles)));
+      }
+
+      return builder.toString();
+    }
+  }
+
+  /**
+   * The max size of the log as derived from the configuration
+   */
+  private static class MaxLogSize {
+    private final int maxSize;
+
+    public MaxLogSize(AccumuloConfiguration config) {
+      maxSize = config.getCount(Property.TABLE_OPERATION_LOG_MAX_SIZE);
+    }
+
+    public int getMaxSize() {
+      return maxSize;
+    }
+  }
+
+  /**
+   * A simple implementation of a ring buffer. Note that this is not thread safe when it comes to
+   * modifications. However the read methods can be done concurrently with a write method.
+   */
+  public static class Ring<T> {
+    private final Object[] ring;
+    private volatile int first;
+    private volatile int last;
+    private static final int overrunThreshold = (Integer.MAX_VALUE / 2);
+
+    public Ring(int size) {
```

Comment:
What happens as the size of the buffer approaches and then surpasses `overrunThreshold`.  For example creating ring buffers like the following.

 * `new Ring(Integer.MAX_VALUE / 2 - 1)`
 * `new Ring(Integer.MAX_VALUE / 2 )`
 * `new Ring(Integer.MAX_VALUE / 2  + 1)`
 * `new Ring(Integer.MAX_VALUE / 2  + 100)`

Wondering if some sort of check need to be done on size like `Preconditions.checkArg(size<overrunThreshold)` but not sure.